### DISCUSSION TITLE
Task Extension at the Backend Interface

### DISCUSSION THREAD

**Discussion by decibelcooper** [cschleiden/go-workflows#285] [General]
I'm struggling to see the vision for the ExtendWorkflowTask and ExtendActivityTask Backend methods.  Today, it looks like these are not yet used by go-workflows.  Additionally, the lock timeout duration is not exposed through the Backend interface, so one would have to have knowledge of the implementation to know what the duration is.

It seems to me that locks durations/deadlines for *Tasks should be something exposed through through the interface, on the *Task itself.  Perhaps per-task timeouts could be defined or hinted by event metadata down the road without further change to the Backend interface.

**Comment by cschleiden** [comment]
The idea is somewhat like this:

```mermaid
sequenceDiagram
    Worker->>+Backend: GetTask
    Backend-->>-Worker: GetTask
    Worker->>+Executor: ExecuteTask
    Note right of Worker: Some time passes, need to extend lock
    Worker->>+Backend: ExtendTask
    Backend-->>-Worker: ExtendTask
    Note right of Worker: Some time passes
    Worker->>+Backend: ExtendTask
    Backend-->>-Worker: ExtendTask
    Executor-->>-Worker: ExecuteTask
    Worker->>+Backend: CompleteTask
    Backend-->>-Worker: CompleteTask
```

If it takes longer than the lock timeout for the executor to finish a task, we need to extend the lock for the task to prevent another worker from work-stealing the task assuming the original worker crashed.

**Reply by decibelcooper** [reply]
Thank you!

I think my concern can be summarized by: _the difficulty or limitation that the Backend interface imposes on determining `need to extend lock`._  

For example, let's say I want to implement a backend that establishes worker membership in a membership ring.  The validity of that membership establishes an implicit lock on tasks for a subset of workflow instances.  In that case, it's not reasonable for each member to independently decide on the lifetime of a lock, but instead a consensus must be reached between workers.

Is it conceivable that the Backend interface could change to accommodate this?  Again, what I mean by that is for tasks to declare their own deadlines, and probably also for ExtendTask to return an updated deadline.
